dll/win32/framedyn/chptrarr.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+* PROJECT:     ReactOS system libraries
+* LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+* PURPOSE:     CHPtrArray class implementation
+* COPYRIGHT:   Copyright 2018 Jin Juanshi (2033537949@qq.com)
+*/
+
+#include <chptrarr.h>
+
+#define GROW_BY_MAX 256
+
+CHPtrArray::CHPtrArray()
+{
+    m_pData = NULL;
+    m_nSize = 0;
+    m_nMaxSize = 0;
+    m_nGrowBy = 0;
+}
+
+CHPtrArray::~CHPtrArray()
+{
+    if (m_pData)
+    {
+        free(m_pData);
+    }
+}
+
+int CHPtrArray::GetSize() const
+{
+    return m_nSize;
+}
+
+int CHPtrArray::GetUpperBound() const
+{
+    return m_nSize - 1;
+}
+
+void CHPtrArray::SetSize(int nNewSize, int nGrowBy) throw (CHeap_Exception)
+{
+    if (nGrowBy != -1)
+    {
+        m_nGrowBy = nGrowBy;
+    }
+
+    if (nNewSize == 0)
+    {
+        if (m_pData)
+        {
+            free(m_pData);
+        }
+        m_pData = NULL;
+        m_nMaxSize = 0;
+        m_nSize = 0;
+        return;
+    }
+
+    if (m_pData == NULL)
+    {
+        m_pData = (void**)malloc(nNewSize * sizeof(void*));
+        m_nMaxSize = nNewSize;
+        m_nSize = nNewSize;
+        return;
+    }
+
+    if (nNewSize <= m_nMaxSize)
+    {
+        if (nNewSize > m_nSize)
+        {
+            memset(&m_pData[m_nSize], 0, (nNewSize - m_nSize) * sizeof(void*));
+            m_nSize = nNewSize;
+            return;
+        }
+    }
+
+    unsigned int uNewGrowBy = m_nGrowBy;
+    if (uNewGrowBy == 0)
+    {
+        uNewGrowBy = m_nSize / 8;
+
+        if (uNewGrowBy < sizeof(void*))
+            uNewGrowBy = sizeof(void*);
+
+        if (uNewGrowBy > GROW_BY_MAX * sizeof(void*))
+            uNewGrowBy = GROW_BY_MAX * sizeof(void*);
+    }
+
+    int nNewMaxSize = m_nMaxSize + uNewGrowBy;
+    if (nNewSize >= nNewMaxSize)
+    {
+        nNewMaxSize = nNewSize;
+    }
+
+    void** pNewData = (void**)malloc(nNewMaxSize * sizeof(void*));
+    memcpy(pNewData, m_pData, m_nSize * sizeof(void*));
+    memset(&pNewData[m_nSize], 0, (nNewSize - m_nSize) * sizeof(void*));
+    free(m_pData);
+
+    m_pData = pNewData;
+    m_nSize = nNewSize;
+    m_nMaxSize = nNewMaxSize;
+}
+
+void CHPtrArray::FreeExtra() throw (CHeap_Exception)
+{
+    if (m_nSize != m_nMaxSize)
+    {
+        void **pNewData = NULL;
+
+        if (m_nSize)
+        {
+            pNewData = (void**)malloc(m_nSize * sizeof(void*));
+            memcpy(pNewData, m_pData, m_nSize * sizeof(void*));
+        }
+
+        if (m_pData)
+        {
+            free(m_pData);
+        }
+        
+        m_pData = pNewData;
+        m_nMaxSize = m_nSize;
+    }
+}
+
+void CHPtrArray::RemoveAll()
+{
+    SetSize(0);
+}
+
+void* CHPtrArray::GetAt(int nIndex) const
+{
+    return m_pData[nIndex];
+}
```

Comment:
I'm quite skeptical regarding `HPtrArray::GetAt()` (same thing for other functions). The function is declared with `void` data type which means it will not return anything upon its termination, yet it does actually return `m_pData[nIndex]` which holds some data.

You cannot return any kind of data, let it be an integer or whatever else as long as your function is set as void. Unless your function is supposed to not return anything then it's fine, period.

Reply:
Note that it's `void*`, not `void`. the asterisk means "pointer to", and in C, "pointer to void" means "pointer to anything".

Reply:
Oh, then that makes sense now. Thanks for clarification.